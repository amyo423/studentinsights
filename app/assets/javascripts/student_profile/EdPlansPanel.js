import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import {toMomentFromRailsDate} from '../helpers/toMoment';


// Renders ed plan data (504)
export default class EdPlansPanel extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      showRaw: false,
      showAll: false
    };
    this.onShowRawClicked = this.onShowRawClicked.bind(this);
    this.onShowMore = this.onShowMore.bind(this);
  }

  onShowRawClicked(e) {
    e.preventDefault();
    this.setState({showRaw: true});
  }

  onShowMore(e) {
    e.preventDefault();
    this.setState({showAll: true});
  }

  render() {
    const {edPlans} = this.props;
    const {showAll, showRaw} = this.state;

    const sortedEdPlans = _.sortBy(edPlans, 'sep_effective_date').reverse();
    const visibleEdPlans = (showAll) ? sortedEdPlans : sortedEdPlans.slice(0, 1);
    return (
      <div className="EdPlansPanel">
        {visibleEdPlans.map(this.renderEdPlan, this)}
        {showAll || edPlans.length <= 1
          ? null
          : <a href="#" style={styles.link} onClick={this.onShowMore}>Show older 504 plans</a>
        }
        {showRaw
          ? <pre>{JSON.stringify(edPlans, null, 2)}</pre>
          : <a href="#" style={styles.link} onClick={this.onShowRawClicked}>Show raw data</a>
        }
      </div>
    );
  }

  // Might not be active or effective
  renderEdPlan(edPlan) {
    const {studentName} = this.props;

    return (
      <div key={edPlan.id} style={{marginBottom: 40, padding: 20, border: '1px solid #ccc', backgroundColor: '#f8f8f8'}}>
        <h2 style={{display: 'flex', flexDirection: 'row', justifyContent: 'space-between'}}>
          <div>504 plan for {studentName}</div>
          <div style={{color: '#ccc', paddingLeft: 10}}>id:{edPlan.sep_oid}</div>
        </h2>
        <div>
          <div style={{paddingTop: 10}}>
            <div>Specific disability: {edPlan.sep_fieldd_006}</div>
            <div>Date of implementation: {reformatDate(edPlan.sep_effective_date)}</div>
            <div>End date: {reformatDate(edPlan.sep_end_date, '(none)')}</div>
            <div>Review date: {reformatDate(edPlan.sep_review_date, '(none)')}</div>
          </div>
          {this.renderSubstanceIfPresent(edPlan)}
          <div style={styles.section}>
            <h6>Persons responsible</h6>
            {reformatPersons(edPlan.sep_fieldd_007, 'None listed.')}
          </div>
          <div style={styles.section}>
            <h6>Accommodations</h6>
            {this.renderAccommodationsList(edPlan)}
          </div>
        </div>
      </div>
    );
  }

  renderSubstanceIfPresent(edPlan) {
    const lines = [
      edPlan.sep_fieldd_001,
      edPlan.sep_fieldd_002,
      edPlan.sep_fieldd_003,
      edPlan.sep_fieldd_004,
      edPlan.sep_fieldd_005,
      // not 6 or 7
      edPlan.sep_fieldd_008,
      edPlan.sep_fieldd_009,
      edPlan.sep_fieldd_010,
      edPlan.sep_fieldd_011,
      edPlan.sep_fieldd_012,
      edPlan.sep_fieldd_013,
      edPlan.sep_fieldd_014
    ];
    const cleanedLines = _.compact(lines.map(field => field === 'N' ? null : field));
    if (cleanedLines.length === 0 || cleanedLines.join('') === '') return null;

    return (
      <div style={styles.section}>
        {cleanedLines.map(line => <div key={line} style={{marginBottom: 10}}>{line}</div>)}
      </div>
    );
  }

  renderAccommodationsList(edPlan) {
    if (edPlan.ed_plan_accommodations.length === 0) return <div>None listed.</div>;

    const filteredAccommodations = edPlan.ed_plan_accommodations.filter(accommodation => {
      return (accommodation.iac_description !== '');
    });
    return (
      <ul style={styles.simpleList}>
        {filteredAccommodations.map(accommodation => {
          const whomEl = accommodation.iac_field
            ? ` (${accommodation.iac_field})`
            : null;
          return <li key={accommodation.id}>{accommodation.iac_description}{whomEl}</li>;
        })}
      </ul>
    );      
  }
}
EdPlansPanel.propTypes = {
  studentName: PropTypes.string.isRequired,
  edPlans: PropTypes.arrayOf(PropTypes.shape({
    sep_effective_date: PropTypes.string.isRequired,
    sep_review_date: PropTypes.string,
    sep_end_date: PropTypes.string,
    ed_plan_accommodations: PropTypes.arrayOf(PropTypes.shape({
      iac_content_area: PropTypes.string,
      iac_category: PropTypes.string,
      iac_type: PropTypes.string,
      iac_description: PropTypes.string,
      iac_field: PropTypes.string,
      iac_name: PropTypes.string,
      iac_last_modified: PropTypes.string
    })).isRequired
  })).isRequired
};

const styles = {
  substance: {
    whiteSpace: 'pre-wrap'
  },
  section: {
    marginTop: 20,
    paddingTop: 10,
    borderTop: '1px solid #ccc'
  },
  simpleList: {
    paddingLeft: 15
  },
  link: {
    display: 'block',
    fontSize: 14,
    marginTop: 10,
    marginBottom: 10
  }
};


function reformatDate(maybeDateText, elseValue = null) {
  if (!maybeDateText) return elseValue;
  return toMomentFromRailsDate(maybeDateText).format('M/D/YYYY');
}

function reformatPersons(personsText, elseValue = null) {
  if (!personsText || personsText === '') return elseValue;
  return (
    <ul style={{...styles.simpleList, listStyleType: 'circle'}}>
      {personsText.split(/[,;]\s/).map(personText => (
        <li key={personText}>{personText}</li>
      ))}
    </ul>
  );
}
