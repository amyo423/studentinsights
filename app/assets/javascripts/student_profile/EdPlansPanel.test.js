import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import EdPlansPanel from './EdPlansPanel';

export function testProps(props = {}) {
  return {
    studentName: 'Garfield Skywalker',
    edPlans: [{
      "id": 245,
      "sep_oid": "fake-sep-oid-1",
      "student_id": 104,
      "sep_grade_level": "N",
      "sep_status": "2",
      "sep_effective_date": "2015-09-15",
      "sep_review_date": "2015-10-15",
      "sep_last_meeting_date": null,
      "sep_district_signed_date": null,
      "sep_parent_signed_date": "2015-09-15",
      "sep_end_date": "2015-10-15",
      "sep_last_modified": null,
      "sep_fieldd_001": "These interventions strategies have been in place since 9/15/15 to 10/15/15: Garfield will be allowed to draw a picture on school assignments.",
      "sep_fieldd_002": "",
      "sep_fieldd_003": "",
      "sep_fieldd_004": "Team Members Present:  Rich Districtwide, Laura Principal.  Email with 504 sent to teachers, guidance counselor, principal on 9/15/15.",
      "sep_fieldd_005": "Category",
      "sep_fieldd_006": "N",
      "sep_fieldd_007": "N",
      "sep_fieldd_008": "N",
      "sep_fieldd_009": "N",
      "sep_fieldd_010": "N",
      "sep_fieldd_011": "N",
      "sep_fieldd_012": "N",
      "sep_fieldd_013": "N",
      "sep_fieldd_014": "N",
      "created_at": "2018-12-08T15:18:53.989Z",
      "updated_at": "2018-12-08T15:18:53.989Z",
      "ed_plan_accommodations": [{
        "id": 601,
        "ed_plan_id": 245,
        "iac_oid": "fixture-oid-1",
        "iac_sep_oid": "fixture-sep-oid-1",
        "iac_content_area": "All",
        "iac_category": "Non-standard",
        "iac_type": "504",
        "iac_description": "Student will meet with teachers before or after school to find out about any make-up work.",
        "iac_field": "",
        "iac_name": "Other Non-Standard Accommodation",
        "iac_last_modified": "2015-03-06T12:43:14.000Z",
        "created_at": "2015-12-09T13:12:20.255Z",
        "updated_at": "2015-12-09T13:12:20.255Z"
      },
      {
        "id": 602,
        "ed_plan_id": 245,
        "iac_oid": "fixture-oid-2",
        "iac_sep_oid": "fixture-sep-oid-1",
        "iac_content_area": "All",
        "iac_category": "Non-standard",
        "iac_type": "504",
        "iac_description": "Prompt the student to review the work they complete before moving on to the next assignment.",
        "iac_field": "",
        "iac_name": "Other Non-Standard Accommodation",
        "iac_last_modified": "2015-03-06T12:43:14.000Z",
        "created_at": "2018-12-09T13:12:20.928Z",
        "updated_at": "2018-12-09T13:12:20.928Z"
      }]
    }, {
      "id": 246,
      "sep_oid": "fake-sep-oid-2",
      "student_id": 104,
      "sep_grade_level": "N",
      "sep_status": "4",
      "sep_effective_date": "2015-08-23",
      "sep_review_date": "2015-10-02",
      "sep_last_meeting_date": null,
      "sep_district_signed_date": null,
      "sep_parent_signed_date": null,
      "sep_end_date": "2016-02-22",
      "sep_last_modified": null,
      "sep_fieldd_001": "",
      "sep_fieldd_002": "",
      "sep_fieldd_003": "",
      "sep_fieldd_004": "",
      "sep_fieldd_005": "",
      "sep_fieldd_006": "ADHDGeneral Anxiety Disorder",
      "sep_fieldd_007": "",
      "sep_fieldd_008": "N",
      "sep_fieldd_009": "N",
      "sep_fieldd_010": "N",
      "sep_fieldd_011": "N",
      "sep_fieldd_012": "N",
      "sep_fieldd_013": "N",
      "sep_fieldd_014": "N",
      "created_at": "2018-12-08T15:18:54.007Z",
      "updated_at": "2018-12-08T15:18:54.007Z",
      "ed_plan_accommodations": []
    }]
  };
}

function testEl(props) {
  return <EdPlansPanel {...props} />;
}

function testRender(props) {
  const el = document.createElement('div');
  ReactDOM.render(testEl(props), el);
  return el;
}

it('renders without crashing', () => {
  testRender(testProps());
});

it('snapshots', () => {
  const props = testProps();
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});