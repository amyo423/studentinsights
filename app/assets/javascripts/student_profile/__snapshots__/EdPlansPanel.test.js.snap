// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="EdPlansPanel"
>
  <div
    style={
      Object {
        "backgroundColor": "#f8f8f8",
        "border": "1px solid #ccc",
        "marginBottom": 40,
        "padding": 20,
      }
    }
  >
    <h2
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <div>
        504 plan for 
        Garfield Skywalker
      </div>
      <div
        style={
          Object {
            "color": "#ccc",
            "paddingLeft": 10,
          }
        }
      >
        id:
        fake-sep-oid-1
      </div>
    </h2>
    <div>
      <div
        style={
          Object {
            "paddingTop": 10,
          }
        }
      >
        <div>
          Specific disability: 
          N
        </div>
        <div>
          Date of implementation: 
          9/15/2015
        </div>
        <div>
          End date: 
          10/15/2015
        </div>
        <div>
          Review date: 
          10/15/2015
        </div>
      </div>
      <div
        style={
          Object {
            "borderTop": "1px solid #ccc",
            "marginTop": 20,
            "paddingTop": 10,
          }
        }
      >
        <div
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          These interventions strategies have been in place since 9/15/15 to 10/15/15: Garfield will be allowed to draw a picture on school assignments.
        </div>
        <div
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          Team Members Present:  Rich Districtwide, Laura Principal.  Email with 504 sent to teachers, guidance counselor, principal on 9/15/15.
        </div>
        <div
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          Category
        </div>
      </div>
      <div
        style={
          Object {
            "borderTop": "1px solid #ccc",
            "marginTop": 20,
            "paddingTop": 10,
          }
        }
      >
        <h6>
          Persons responsible
        </h6>
        <ul
          style={
            Object {
              "listStyleType": "lower-alpha",
              "paddingLeft": 15,
            }
          }
        >
          <li>
            N
          </li>
        </ul>
      </div>
      <div
        style={
          Object {
            "borderTop": "1px solid #ccc",
            "marginTop": 20,
            "paddingTop": 10,
          }
        }
      >
        <h6>
          Accommodations
        </h6>
        <ul
          style={
            Object {
              "listStyleType": "lower-alpha",
              "paddingLeft": 15,
            }
          }
        >
          <li>
            Student will meet with teachers before or after school to find out about any make-up work.
          </li>
          <li>
            Prompt the student to review the work they complete before moving on to the next assignment.
          </li>
        </ul>
      </div>
    </div>
  </div>
  <a
    href="#"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
        "fontSize": 14,
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    Show older 504 plans
  </a>
  <a
    href="#"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
        "fontSize": 14,
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    Show raw data
  </a>
</div>
`;
